Add support for WCH CH32V303

miggazElquez opened the pull request:
This PR add support for the SoC CH32V303 from WCH, and for the board CH32V303EVT.

Depends on #85395 for the support of `west flash`.

miggazElquez committed:
soc: Add support for the WCH CH32V303

Adds support for building an image for the ch32v303.

Signed-off-by: Miguel Gazquez <miguel.gazquez@bootlin.com>

miggazElquez committed:
boards: Add support for the board CH32V303EVT

Adds support for the CH32V303EVT board, based on the CH32V303 soc.

Signed-off-by: Miguel Gazquez <miguel.gazquez@bootlin.com>

miggazElquez committed:
modules: Update hal_wch

Update hal_wch.

As the hal upstream changed name, there is now a name conflict.
Rename ch32fun.h to hal_ch32fun.h to fix this conflict.

Signed-off-by: Miguel Gazquez <miguel.gazquez@bootlin.com>

miggazElquez commented:
@VynDragon if you want to take a look, as this is based on your code for the ch32v208.

VynDragon commented:
Looks perfect to me as far as I can tell beside the nordicjm remarks.
CONFIG_FLASH_BASE_ADDRESS shouldnt be set like this in there unless you want to enable extended code flash, but since it's set to 128K it can just stay unset/0 . What you could do instead is add a second flash zone at 0x0800 0000 + 128KB of size 480KB - 128KB

miggazElquez commented on boards/wch/ch32v303evt/ch32v303evt_defconfig:
```diff
@@ -0,0 +1,10 @@
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+CONFIG_FLASH_BASE_ADDRESS=0x08000000
```
If I don't put that, `CONFIG_FLASH_BASE_ADDRESS` will be set to the address of the flash in the device tree (currently 0), and minichlink fail to flash at the address 0. But if I change the address in the device tree, I can flash but Zephyr then fails to boot, I don't really know why. @VynDragon you didn't have this issue on the 208 ?

What would be the way to solve this ? I saw that some boards did change `CONFIG_FLASH_BASE_ADDRESS` in their defconfig, so that's why I put it there, but maybe it should be in another place ?

VynDragon replied:
Oh so that's why the west flash didnt work! Ill look into it, might be a matter of fixing the runner instead of this, otherwise figuring how to make it boot in non aliased space would make it easier to use the full flash so ill see for both issues.

VynDragon replied:
Well it works for me when i set 			flash0: flash@8000000 {
				compatible = "soc-nv-flash";
				reg = <0x08000000 DT_SIZE_K(128)>;
			};
			
			But west flash still doenst work so no clue what's going on there

VynDragon replied:
a it's the -b for the runner, need to disable reset for ch32v208

VynDragon replied:
Annnnd it's force enabled because zephyr doesnt believe in not resetting every board after flashing for some reason (in run_common.py)

VynDragon replied:
> Well it works for me when i set
> 
> ```
> flash0: flash@8000000 {
> 				compatible = "soc-nv-flash";
> 				reg = <0x08000000 DT_SIZE_K(128)>;
> 			};
> ```
> 
> after udpating minichlink and using `minichlink -w build/zephyr/zephyr.bin 0x0` But west flash still doenst work so no clue what's going on there

NVM i got tricket by flash vs ram.

VynDragon replied:
Ok here, this has everything: https://github.com/VynDragon/zephyr/tree/ch32v208_extended_flash_demo
Full run on flash and a swath of random fixes for west and minichlink. You need latest minichlink that allows flashing over 128KB.

VynDragon replied:
Oh also if you wonder like i did why mtvec 31:2 was 0 on ch32v003 instead of 1 like it hsould have been with the first jump, it's because it needs to be 1024 bytes aligned on ch32v003. on ch32v20x and 30x it's fine to have it pointing where it should (so 31:2 points to 3x4 bytes 0x3 << 2 address with the full flash boot and 1x4 0x1 << 2 for regular mode, which i got wrong on ch32v208 (it's okay though because it still works unless the window watchdog interrupt is called then it resets the cpu)).

VynDragon replied:
It might be a frequency issue, for some reason on CH32V208 it works at 144MHz, but they 'highly recommend' using 120M in the RM for the F/V20x/30x serie.


> For flashing

Me neither, it's supposed to flash to 0x08000000 so with dt-flash=y and the flash address set to that and the long jump code, or 0x0 with no long jump and 0x08000000 flash address still. I think it changed between old and new minichlink version?

miggazElquez replied:
So before answering to you I wanted to test it one last time and now it works, I don't know what I did wrong during my tests before... because I'm quite sure I tested with the long jump, but maybe I had changed something else, I don't know. Anyway it works now, so I modified my PR to include it, it seems cleaner to me than passing the flash address directly to minichlink

VynDragon replied:
> I still don't really understand the point of the long jump

It's so the pc is in the right zone when the short jumps happen, eg that the pc is in the area expected when it tries to do a short jump, there is also the matter of the gp pointer, and a couple other thing i believe break when operated in not the expected area.  In practice it crashes if you run it without that long jump.

> Because when testing it seems like we boot in non-aliased mode

It always boot in the aliased zone?

miggazElquez replied:
> It always boot in the aliased zone?

I think that I was fooled by the debugger, I'm not sure about what is happening but with the debugger even without the longjump I'm in non aliased space (pc = 0x0800...). I think it's not really executing the first instructions, it seems like it's just jumping to __start directly ?

VynDragon replied:
Weird, I dont know what could be going on

nzmichaelh commented on boards/wch/ch32v303evt/doc/index.rst:
```diff
@@ -0,0 +1,58 @@
+
+Applications for the ``ch32v303evt`` board target can be built and flashed
+in the usual way (see :ref:`build_an_application` and :ref:`application_run`
+for more details); however, an external programme(like) is required since the board
```
"an external programmer"?

miggazElquez replied:
good catch, thanks !

nzmichaelh commented on west.yml:
```diff
@@ -258,7 +258,7 @@ manifest:
         - hal
     - name: hal_wch
-      revision: 1de9d3e406726702ce7cfc504509a02ecc463554
+      revision: pull/1/head
```
Is this correct?

miggazElquez replied:
I think, it's to indicate that this PR depend on the PR \#1 on the hal_wch repository. The message from zephyrbot at the top of the PR looks good so I think it's correct.

nzmichaelh reviewed (dismissed):
LGTM

miggazElquez commented on soc/wch/ch32v/qingke_v4f/vector.S:
```diff
@@ -0,0 +1,29 @@
+	.endr
+
+SECTION_FUNC(vectors, __start)
+	li 	a0, 3
```
I suppose we should fix this for the 208 also ? Even it doesn't really create any problem.

VynDragon commented:
I cant find the other comment so here the PR for fixing CH32V208 with the issue mentionned: https://github.com/zephyrproject-rtos/zephyr/pull/87345, inherently fixing it by adding the full flash feature

miggazElquez commented on soc/wch/ch32v/qingke_v4f/vector.S:
```diff
@@ -0,0 +1,31 @@
+SECTION_FUNC(vectors, __start)
+	li 	a0, 0xf
+	csrw	mtvec, a0
+	j	__initialize
```
I think it's the same for the ch32v208 (Qingke V4C), but not for the ch32v003(Qingke V2A) (see https://github.com/zephyrproject-rtos/zephyr/pull/87125#discussion_r1996359739).

Should I still de-duplicate it ?

VynDragon replied:
Agreed yea, this version is only relevant for CH32V208, and CH32V30x (unless i'm missing some)

fkokosinski replied:
@miggazElquez yeah, let's de-duplicate it :)

miggazElquez replied:
yeah, with the new Pr introducing the v203 it seems even more logical.

As the new vector.S use the non-aliased space, it needs some changes in the devicetree of the 208 to use the new files, so I didn't remove it from here. These changes are in #87345, which updated the SoC to use the new version of `vector.S`.

VynDragon replied:
Should i submit the deduplication with the 208 update? Should it be qingke_v4/f, qingke_v4/b, etc, or stay qingke_v4f, qingke_v4c, with a additional qingke_v4 ?

fkokosinski replied:
The latter sounds better, and it replicates the convention used for other SoCs (stm, gd, silabs), but I'd rearrange it to something like this:
```
- chv32 (family)
  - qingke_v4 (series)
    - common
    - qingke_v4f (soc)
    - qingke_v4c (soc)
```

fkokosinski replied:
We can handle it in the update IMO

VynDragon replied:
Ok will update that PR

SoftwareArchitekt commented on boards/wch/ch32v303evt/doc/index.rst:
```diff
@@ -0,0 +1,59 @@
+.. zephyr:board:: ch32v303evt
```
Here it is not clear, which board is mentioned. WCH have several CH32V303-Boards. They differ in size of flash and if FSMC is there or not. So please specify more clear which type is described (VH32V303CBT6 or VH32V303VCT6 or VH32V303RCT6)

miggazElquez replied:
Ah I hadn't noticed that there are multiple development kit named CH32V303EVT. Mine is for the one with the CH32V303VCT6. I'm not sure about the best way to handle it : I can of course add it to the text of the documentation, but for the name of the board I'm not sure of the correct way t handle it (in this line of the documentation but also in general, the name of the board is used by west). Maybe this where I have to use board variant ? I'm not sure

SoftwareArchitekt replied:
I am actually working on the 308. There I have the same problem with different naming. Here I guess, we need to add the extension to the naming. I future there will then be 3 boards. If you look here https://www.wch-ic.com/products/CH32V303.html in part Product Selection Guide you can see a lot of diffenerces (number of adc, Uart SPI etc, Flashsize Ram-Size). So my proposale is to add to yout board name the processor extension like WCH to it at the order of dev-boards too.

SoftwareArchitekt replied:
thats looks fine from myside

SoftwareArchitekt replied:
I will look in the next days over all your files and compare it with my 308-files (so far i have them) to check their compatibility

miggazElquez commented:
I modified the commit for the hal update, and renamed the board.

Please put a DNM tag on this PR, I'm still waiting for #87345 to be merged before I can rebase my code.

SoftwareArchitekt commented on modules/hal_wch/hal_ch32fun.h:
```diff
@@ -9,12 +9,21 @@
 
 #if defined(CONFIG_SOC_SERIES_QINGKE_V4C)
 #define CH32V20x 1
-#include <ch32v003fun.h>
```
i am not happy with the filename starting with wch. I have a board using the Qingke-V4F and not from wch. I guess, that the filename should be change to qingkeCh32fun or something like this.

SoftwareArchitekt replied:
Sorry for confusion, I see now, that the chip is still from wch. So evrything can stay.

SoftwareArchitekt commented on soc/wch/ch32v/qingke_v4_vector.S:
I guess this file should also be located in qinke_v4f-directory

miggazElquez replied:
For this file I'm waiting on #87345 to be merged to rebase my code, so this file will end up in a folder for all Qingke-V4

SoftwareArchitekt commented on soc/wch/ch32v/qingke_v4f/pinctrl_soc.h:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @brief Type to hold a pin's pinctrl configuration.
+ */
+struct ch32v303_pinctrl_soc_pin {
```
this struct will also be used by ch32v307. This means the naming would be better qingke_v4f_pin....

miggazElquez replied:
Same for this, it's in the folder for all qingke v4

SoftwareArchitekt replied:
as you told, it is in folder for all qinke v4, so a name of a soc (ch32v303) is maybe wrong. better to use a struct qinke_v4_pinvtrl_soc_pin {

miggazElquez replied:
So the PR deduplicating the files was changed to just focus on the changes for the v208, so I created another PR deduplicating all the files, with the idea that when one of these two PR is merged, I would just rebase the other.

About the specific point, for the pinctrl file, naming it qingke_v4_something is maybe not the best idea, see [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/90007#discussion_r2094022808).

nzmichaelh replied:
+1 to not naming it after the microarchitecture, as the pinctrl is a feature of the SoC, not the processor.

miggazElquez commented:
The deduplication of the files is now in #90007; when one of these is merged I will rebase the other.

nzmichaelh commented:
I've sent https://github.com/zephyrproject-rtos/zephyr/pull/90438 to fix the other drivers after the ch32fun -> hal_ch32fun update.
